### ISSUE TITLE
Allow zero-method interfaces

### ISSUE THREAD

**Issue by axch** [google-research/dex-lang#1055]
It's occasionally handy to define interfaces with no methods.  For instance,
```
interface [Group g] AbelianGroup g
```
would serve as a marker that on the type `g`, the operation defined by `Group` in fact commutes.

Another use is as an interface alias.  For example, defining
```
interface [VSpace f, Fractional f, Mul f, Add f] MyGoodType f
instance {f} [VSpace f, Fractional f, Mul f, Add f] MyGoodType f
```
saves a fair amount of typing in situations where that particular combination of class constraints arises frequently.

Either of these use-cases can of course be faked by defining the interface to require a trivially implemented method with a dummy name, but it would be nice if we didn't need that workaround.

**Comment by duvenaud** [comment]
Another example where this would be useful is for the `NonEmpty` typeclass, whose implementations are all identical and could be moved outside:

```
instance NonEmpty Unit
  first_ix = unsafe_from_ordinal _ 0

instance {a b} [NonEmpty a, NonEmpty b] NonEmpty (a & b)
  first_ix = unsafe_from_ordinal _ 0

instance {a b} [Ix b, NonEmpty a] NonEmpty (a|b)
  first_ix = unsafe_from_ordinal _ 0

instance {a b} [Ix a, NonEmpty b] NonEmpty (a|b)
  first_ix = unsafe_from_ordinal _ 0

instance {a} [Ix a] NonEmpty (Maybe a)
  first_ix = unsafe_from_ordinal _ 0
```

**Comment by axch** [comment]
Not to distract from the value of implementing the feature, but in the case of `NonEmpty` I actually think a method is appropriate: the method gives the proof that the set is, actually, not empty.  `NonEmpty` is an excellent use-case for default method definitions, however.

**Comment by duvenaud** [comment]
I would agree with you if the function actually did provide a guarantee, but because it's using `unsafe_from_ordinal`, it typechecks even for empty tables:

```
instance NonEmpty (Fin 0)
  first_ix = unsafe_from_ordinal _ 0

a : (Fin 0) = last_ix
a
(4294967295@(Fin 0))
```

So in this instance, I don't think the method is a proof of anything.

**Comment by axch** [comment]
I think you're saying that the definitions
```
interface NonEmpty a

def first_ix {n} [NonEmpty n] : n = unsafe_from_ordinal _ 0
```
are equivalent to the status quo.  In this sense I agree -- implementing the class unsafely doesn't prove anything.  I was arguing that we want the `NonEmpty` constraint on `first_ix` one way or another.

**Comment by duvenaud** [comment]
Oh, yes, I agree that if we moved `first_ix` out of the interface then it should still have the `NonEmpty` constraint, just like `last_ix` does.

**Comment by duvenaud** [comment]
I just realized that there is a related useful feature of zero-method interfaces:  There's no need to have a unique derivation for any given instance.

This is already a problem for `NonEmpty`.  For instance, the type `(Unit | Unit)` can be proven `NonEmpty` by both the left and the right sides, giving this error:
```
Type error:Multiple candidate class dictionaries for: (NonEmpty ((Unit | Unit))
```

So if we add this feature we should also turn off the uniqueness check when the dictionaries are empty.

**Comment by dougalm** [comment]
I think this is mainly a concrete syntax problem. We don't require a non-empty list of methods anywhere else.

It should be easy to fix. Our indented blocks are parsed with `ConcreteSyntax.onePerLine` which requires at least one item because it needs to figure out the indentation level. Let's just take a leaf out of Python's book (rather, a lexeme out of Python's lexicon) and add a `pass` lexeme whose only purpose is to let you write blocks that cause `onePerLine` to parse the empty list.

@normanrink may be working on this already.

**Comment by dougalm** [comment]
Fixed by #1241
